### PR TITLE
Circular date representations to default date transmogrifier

### PR DESCRIPTION
**Pull request by leahmcguire**
**Related issues**
Adding circular date representations to default date transmogrifier

**Describe the proposed solution**
Added a DateMap circular transformer
Make Date circular transformer take a sequence instead of a single feature
Updated and added shortcuts for all Date and DateMap classes toUnitCircle and added default transformations to vectorize for all of these classes
Fixed metadata on Geolocation and GeolocationMap so that keep the name of the column in descriptorValue
 
**Additional context**
Current default is time ago from now but circular representations are better for capturing many business use cases

### CODE REVIEW

File: core/src/main/scala/com/salesforce/op/dsl/RichDateFeature.scala [hunk 1]
```diff
@@ -85,17 +91,22 @@ trait RichDateFeature {
...
+     * @param circularDateRepresentations list of all the circular date representations that should be included
+     *                                    feature vector
      * @return result feature of type Vector
      */
     def vectorize
     (
       dateListPivot: DateListPivot,
       referenceDate: JDateTime = TransmogrifierDefaults.ReferenceDate,
       trackNulls: Boolean = TransmogrifierDefaults.TrackNulls,
+      circularDateRepresentations: Seq[TimePeriod] = TransmogrifierDefaults.CircularDateRepresentations,
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/dsl/RichDateFeature.scala]: this is a ridiculously long param name ;) can we perhaps shorten to `circularDateRep`? (same elsewhere)

File: core/src/main/scala/com/salesforce/op/dsl/RichDateFeature.scala [hunk 2]
```diff
@@ -85,17 +91,22 @@ trait RichDateFeature {
...
       others: Array[FeatureLike[Date]] = Array.empty
     ): FeatureLike[OPVector] = {
-      f.toDateList().vectorize(dateListPivot = dateListPivot, referenceDate = referenceDate, trackNulls = trackNulls,
-        others = others.map(_.toDateList()))
+      val timePeriods = circularDateRepresentations.map(tp => f.toUnitCircle(tp, others))
+      val time = f.toDateList().vectorize(dateListPivot = dateListPivot, referenceDate = referenceDate,
+        trackNulls = trackNulls, others = others.map(_.toDateList()))
+      (timePeriods :+ time).combine()
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/dsl/RichDateFeature.scala]: `if (timePeriods.isEmpty) time else (timePeriods :+ time).combine()` (same elsewhere)

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala [hunk 3]
```diff
@@ -30,22 +30,39 @@
...
+  final def setTimePeriod(value: TimePeriod): this.type = set(timePeriod, value.entryName)
+}
+
 /**
  * Following: http://webspace.ship.edu/pgmarr/geo441/lectures/lec%2016%20-%20directional%20statistics.pdf
  * Transforms a Date or DateTime field into a cartesian coordinate representation
  * of an extracted time period on the unit circle
  *
- * @param timePeriod The time period to extract from the timestamp
+ * parameter timePeriod The time period to extract from the timestamp
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: do you mean spark parameter?

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala [hunk 4]
```diff
@@ -60,31 +77,122 @@ import org.joda.time.{DateTime => JDateTime}
...
+ * angle radians = 2*math.Pi*6/24
+ * We return the cartesian coordinates of this point: (math.cos(radians), math.sin(radians))
+ *
+ * The first time period always has angle 0.
+ *
+ * Note: We use the ISO week date format https://en.wikipedia.org/wiki/ISO_week_date#First_week
+ * Monday is the first day of the week
+ * & the first week of the year is the week wit the first Monday after Jan 1.
+ */
+class DateMapToUnitCircleVectorizer[T <: DateMap]
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: please put this class into a separate file.

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala [hunk 5]
```diff
@@ -60,31 +77,122 @@ import org.joda.time.{DateTime => JDateTime}
...
+      descriptorValue = Option(dec)
+    )
 
-  override def transformFn: T => OPVector = timestamp => {
-    val datetime: Option[JDateTime] = timestamp.value.map(new JDateTime(_))
-    val (timePeriod, periodSize) = getTimePeriod match {
+    meta.withColumns(cols.toArray)
+  }
+
+  override def fitFn(dataset: Dataset[Seq[Map[String, Long]]]): SequenceModel[T, OPVector] = {
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: `Dataset[Seq[T#Value]]`

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala [hunk 6]
```diff
@@ -60,31 +77,122 @@ import org.joda.time.{DateTime => JDateTime}
...
+    val meta = makeVectorMetadata(allKeys)
+    setMetadata(meta.toMetadata)
+    new DateMapToUnitCircleVectorizerModel[T](allKeys = allKeys, shouldClean = shouldClean,
+      timePeriodString = getTimePeriod.entryName, operationName = operationName, uid = uid
+    )
+  }
+
+}
+
+final class DateMapToUnitCircleVectorizerModel[T <: DateMap] private[op]
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: docs please

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala [hunk 7]
```diff
@@ -60,31 +77,122 @@ import org.joda.time.{DateTime => JDateTime}
...
+(
+  val allKeys: Seq[Seq[String]],
+  val shouldClean: Boolean,
+  val timePeriodString: String,
+  operationName: String,
+  uid: String
+)(implicit tti: TypeTag[T]) extends SequenceModel[T, OPVector](operationName = operationName, uid = uid)
+  with CleanTextMapFun {
+
+  private val timePeriod: TimePeriod = TimePeriod.withNameInsensitive(timePeriodString)
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: please move `TimePeriod` enum to `features` subproject and register to `OpPipelineStageReadWriteShared.formats`, then you will be able to pass `val timePeriod: TimePeriod` as ctor argument for `DateMapToUnitCircleVectorizerModel`. (see how it's done with `HashSpaceStrategy`)

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala [hunk 8]
```diff
@@ -96,16 +204,8 @@ class DateToUnitCircleTransformer[T <: Date]
...
-  }
-
-  override def onGetMetadata(): Unit = {
-    super.onGetMetadata()
-    val columns = Seq(s"x_${$(timePeriod)}", s"y_${$(timePeriod)}")
-      .map{ iv => in1.toColumnMetaData().copy(indicatorValue = Option(iv)) }.toArray
-    val history = Map(in1.name -> FeatureHistory(originFeatures = in1.originFeatures, stages = in1.stages))
-    setMetadata(OpVectorMetadata(getOutputFeatureName, columns, history).toMetadata)
+    radians.map(r => Array(math.cos(r), math.sin(r))).
+      getOrElse(Array(0.0, 0.0))
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: please reformat: redundant endline.

File: core/src/test/scala/com/salesforce/op/stages/impl/feature/DateMapToUnitCircleVectorizerTest.scala [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+      .fit(inputData).transform(inputData)
+    val actual = transformed.collect(output)
+    all (actual.zip(expectedResult).map(g => Vectors.sqdist(g._1.value, g._2.value))) should be < eps
+  }
+
+  it should "make the correct metadata" in {
+    val fitted = estimator.fit(inputData)
+    val meta = OpVectorMetadata(fitted.getOutputFeatureName, fitted.getMetadata())
+    meta.columns.length shouldBe 4
+    meta.columns.map(_.grouping.get) shouldEqual Seq("a", "a", "b", "b")
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/stages/impl/feature/DateMapToUnitCircleVectorizerTest.scala]: `meta.columns.flatMap(_.grouping)` and `meta.columns.flatMap(_.descriptorValue)`

File: core/src/test/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformerTest.scala [hunk 10]
```diff
@@ -99,8 +99,8 @@ class DateToUnitCircleTransformerTest extends OpTransformerSpec[OPVector, DateTo
     val transformed = vectorizer.transform(ds)
     val meta = OpVectorMetadata(transformed.schema(vectorizer.getOutput().name))
     meta.columns.length should equal (2)
-    meta.columns(0).indicatorValue.get should equal("x_HourOfDay")
-    meta.columns(1).indicatorValue.get should equal("y_HourOfDay")
+    meta.columns(0).descriptorValue.get should equal("x_HourOfDay")
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformerTest.scala]: `meta.columns(0).descriptorValue should be Some("x_HourOfDay")`

File: core/src/test/scala/com/salesforce/op/stages/impl/feature/OPMapVectorizerTest.scala [hunk 11]
```diff
@@ -46,6 +46,7 @@ import org.junit.runner.RunWith
 import org.scalatest.junit.JUnitRunner
 import org.scalatest.{FlatSpec, Matchers}
 import org.slf4j.LoggerFactory
+// import org.apache.log4j.Level
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/stages/impl/feature/OPMapVectorizerTest.scala]: redundant import

File: features/src/main/scala/com/salesforce/op/features/types/package.scala [hunk 12]
```diff
@@ -196,6 +196,13 @@ package object types extends FeatureTypeSparkConverters {
   implicit class VectorConversions(val v: Vector) extends AnyVal {
     def toOPVector: OPVector = new OPVector(v)
   }
+  // Arrays
+  implicit class ArrayDoubleConversions(val v: Array[Double]) extends AnyVal {
+    def toReal: Seq[Real] = v.map(_.toReal)
```
Comment (tovbinm) [inline features/src/main/scala/com/salesforce/op/features/types/package.scala]: keep the Array type in return value, ie. `def toReal: Array[Real]` and `def toRealNN: Array[RealNN]`

**Comment by tovbinm** [comment]
@leahmcguire can you please update the description of the PR shortcuts & vectorizers were added (we will use it in our release notes later).

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala (hunk 3 above)
Comment (leahmcguire) [reply core/src/main/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformer.scala]: yes - it bugged me that it was highlighted red

**Pull request merged**